design/StableRegions-20230209.md

```diff
@@ -0,0 +1,222 @@
+
+   - NB: The organization of this table is not useful for efficient
+     access calculations of load or store (they require a linear
+     search that would be prohibitively slow). Rather, the
+     `region-blocks` table is designed for that purpose.  OTOH, this
+     table is suitable to do a "batch" rebuild of the dynamic heap-allocated vectors
+     in that table, which get lost during upgrades.
+
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BlockRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding block ID.
+
+### region-blocks table
+
+ - purpose: 
+   - relate a region ID to its vector of block IDs, to compute an access efficiently (load or store).
+
+   - NB: The organization of this table is meant to support O(1) load
+     and store operations, but in so doing, it needs to use a
+     dynamically-sized vector for each region.  There is no a priori
+     way to know how to allocate these, and if we naively preallocate
+     them to each be the potential "maximal region" (with all blocks
+     allocated to it), the resulting preallocated table requires
+     gigabytes of space, and thus is prohibitively large.
+
+     At the same time, having dynamically-sized stable vectors is kind
+     of the point of regions (each is such a vector, in a sense), and
+     so requiring these first to implement regions' meta data seems
+     potentially circular in concept, if not also very complex.
+
+     So, we seem forced to use dynamically-sized heap vectors, and to
+     deal the compilations of integrating them with GC (how involved is
+     that?) and rebuilding them on upgrade.
+
+
+ - 32768 entries (statically sized, ignoring the `vec_ptr` objects referenced therein).
+ - 16 bytes per entry.
+ - entry type = `RegionBlocks { size_in_pages: Nat64; vec_capacity : Nat32; vec_ptr: Nat32 }`
+ - the location of each entry in the table gives its corresponding region ID.
+ - the `size_in_pages` field measures the size of the region in terms of _pages_, not blocks. 
+ - the `vec_capacity` and `vec_ptr` fields work with `size_in_pages`
+   to represent a growable vector that we call the region's **"access
+   vector"** (because "blocks vector" sounds a bit strange, and its
+   used to support O(1) access operations): 
+   - the access vector's address is held in `vec_ptr` and it has `vec_capacity` slots.
+   - the first `size_in_pages / 128` slots of `vec_ptr` contain a valid page block ID for the region.
+   - to support dynamic growth, _the access vector is held in the Motoko **heap** rather than **stable memory**._
+   - the access vector doubles when it grows.
+   - no region has more than 32k page blocks, so a `Nat16` suffices for `capacity`,
+   - but we use a `Nat32` for `capacity` to make all fields things word-aligned (does that matter?).
+   - during an upgrade, the access vectors are "rebuilt" in a batch (see `rebuild`).
```

Comment:
I think the simple way to do this would be to just do this during deserialization of the stateful region id. If you actually serialize and deserialize the enclosed blob and pagecount, there is almost nothing extra to do and the extension of the serializer/deserialiser just need an extra case for objects tagged as regions, which need to be serializad /desserialzed in a way that preserves identity, as we already do for mutable arrays and mutable fields.

Reply:
> I think the simple way to do this would be to just do this during deserialization of the stateful region id. 

The current document specifies that values of type `Region` are only an ID with at least 16 bits.  (So their serialization/deserialization is just like a `Nat16`.)

They are not stateful objects, and need not even be heap-allocated (they fit into a `Value` type in the RTS code).

Also, there's no need to serialize any of this table at all -- except for the blobs that implement the "access vectors", it's all just kept in stable memory.

Finally, if we do serialize the access vectors, we'll be serializing data needlessly, since we can reconstruct it from the `block-region` table, which is the ground truth for that relation across updates

Reply:
Your reply predates this discussion on slack https://dfinity.slack.com/archives/CPL67E7MX/p1676502024805799?thread_ts=1675985996.173629&cid=CPL67E7MX, right? Just want to check that I don't need to reply and we are in agreement now (using stateful heap object to represent region etc).

Reply:
Yes, the slack thread is the latest.  Sorry for the confusion.

I just pushed a commit to bring the doc into consistency with our Slack conversation, hopefully.